(Unified)PDFPlugin: "Open" and "Save" write useless incomplete files when the document is not completely loaded

hortont424 opened the pull request:
#### bf47c375c189dccc7b88509bbb997e6e0363c18c
<pre>
(Unified)PDFPlugin: &quot;Open&quot; and &quot;Save&quot; write useless incomplete files when the document is not completely loaded
<a href="https://bugs.webkit.org/show_bug.cgi?id=271497">https://bugs.webkit.org/show_bug.cgi?id=271497</a>
&lt;<a href="https://rdar.apple.com/problem/117451408">rdar://problem/117451408</a>&gt;

Reviewed by Abrar Rahman Protyasha and Simon Fraser.

If Open or Save are clicked before the document is loaded, defer the action
until it is.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::teardown):
Call save/open with invalid data when tearing down the plugin.

(WebKit::PDFPluginBase::streamDidFinishLoading):
Call save/open once we&apos;re actually ready.

(WebKit::PDFPluginBase::save):
(WebKit::PDFPluginBase::openWithPreview):
If called multiple times, let go of the old completion handler by calling it
back with invalid data; only store the most recent one.

Canonical link: <a href="https://commits.webkit.org/276584@main">https://commits.webkit.org/276584@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eabf12ead9c89a2bd676f0fe59fef0d3f0045fda

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45060 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24172 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47564 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47719 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41069 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28261 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21569 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36966 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45638 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21222 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/38805 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18061 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/18636 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3108 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41318 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40234 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49410 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/20034 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16559 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/43985 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21349 "Built successfully") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/42766 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10020 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/21696 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21028 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -403,6 +414,12 @@
 
     tryRunScriptsInPDFDocument();
+
+    if (auto existingCompletionHandler = std::exchange(m_pendingSaveCompletionHandler, nullptr))
```
Consistency 😛 
```suggestion
    if (auto existingCompletionHandler = std::exchange(m_pendingSaveCompletionHandler, { }))
```

hortont424 replied:
this one is _inexplicable_

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -403,6 +414,12 @@
     }
 
     tryRunScriptsInPDFDocument();
+
```
```suggestion

#if ENABLE(PDF_HUD)
```

hortont424 committed:
(Unified)PDFPlugin: "Open" and "Save" write useless incomplete files when the document is not completely loaded
https://bugs.webkit.org/show_bug.cgi?id=271497
<rdar://problem/117451408>

Reviewed by Abrar Rahman Protyasha and Simon Fraser.

If Open or Save are clicked before the document is loaded, defer the action
until it is.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::teardown):
Call save/open with invalid data when tearing down the plugin.

(WebKit::PDFPluginBase::streamDidFinishLoading):
Call save/open once we're actually ready.

(WebKit::PDFPluginBase::save):
(WebKit::PDFPluginBase::openWithPreview):
If called multiple times, let go of the old completion handler by calling it
back with invalid data; only store the most recent one.

Canonical link: https://commits.webkit.org/276584@main
